### PR TITLE
Move slot cleanup to AccountsBackgroundService

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Working towards: https://github.com/solana-labs/solana/pull/13140

Bank drop()->purge_slot() can race with AccountsBackgroundService flushing the cache.

#### Summary of Changes
Send slots to be dropped to AccountsBackgroundService on Bank::drop()

Fixes #

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -4487,9 +4507,16 @@ impl Bank {
 
 impl Drop for Bank {
     fn drop(&mut self) {
-        // For root slots this is a noop
         if !self.skip_drop.load(Relaxed) {
-            self.rc.accounts.purge_slot(self.slot());
+            if let Some(drop_callback) = self.drop_callback.read().unwrap().as_ref() {
```
Comment (carllin) [inline runtime/src/bank.rs:4511]: Crux of sending the AccountsBackgroundService dropped slots

File: core/src/tvu.rs [hunk 2]
```diff
@@ -208,6 +212,22 @@ impl Tvu {
...
 
+        let (pruned_banks_sender, pruned_banks_receiver) = unbounded();
+
+        // Before replay starts, set the callbacks in each of the banks in BankForks
+        for bank in bank_forks.read().unwrap().banks.values() {
+            bank.set_callback(Some(Box::new(SendDroppedBankCallback::new(
+                pruned_banks_sender.clone(),
+            ))));
+        }
+
```
Comment (carllin) [inline core/src/tvu.rs:223]: Set callback in each of the banks in BankForks after blockstore replay is done.

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -701,13 +701,28 @@ impl fmt::Display for RewardType {
...
     pub post_balance: u64, // Account balance in lamports after `lamports` was applied
 }
 
+#[derive(Debug, Default)]
+pub struct OptionalDropCallback(Option<Box<dyn DropCallback + Send + Sync>>);
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+impl AbiExample for OptionalDropCallback {
+    fn example() -> Self {
+        Self(None)
```
Comment (carllin) [inline runtime/src/bank.rs:722]: @ryoqun thanks for the recommendation here!

**Review by sakridge — DISMISSED** [review]
lgtm

**Pull request merged**